internal/controllers/cluster/cluster_controller_phases.go

```diff
@@ -234,27 +245,35 @@ func (r *Reconciler) reconcileInfrastructure(ctx context.Context, s *scope) (ctr
 		return ctrl.Result{}, nil
 	}
 
-	// If the infrastructure provider is not ready (and it wasn't ready before), return early.
-	if !ready && !cluster.Status.InfrastructureReady {
+	// If the InfrastructureCluster is not provisioned (and it wasn't already provisioned before), return.
+	if !*provisioned && !cluster.Status.InfrastructureReady {
 		log.V(3).Info("Infrastructure provider is not ready yet")
 		return ctrl.Result{}, nil
 	}
 
 	// Get and parse Spec.ControlPlaneEndpoint field from the infrastructure provider.
 	if !cluster.Spec.ControlPlaneEndpoint.IsValid() {
-		if err := util.UnstructuredUnmarshalField(s.infraCluster, &cluster.Spec.ControlPlaneEndpoint, "spec", "controlPlaneEndpoint"); err != nil && err != util.ErrUnstructuredFieldNotFound {
-			return ctrl.Result{}, errors.Wrapf(err, "failed to retrieve Spec.ControlPlaneEndpoint from infrastructure provider for Cluster %q in namespace %q",
-				cluster.Name, cluster.Namespace)
+		if endpoint, err := contract.InfrastructureCluster().ControlPlaneEndpoint().Get(obj); err == nil {
+			cluster.Spec.ControlPlaneEndpoint = *endpoint
+		} else {
+			if !errors.Is(err, contract.ErrFieldNotFound) {
+				return ctrl.Result{}, errors.Wrapf(err, "failed to retrieve %s from infrastructure provider for Cluster %q in namespace %q",
+					strings.Join(contract.InfrastructureCluster().ControlPlaneEndpoint().Path(), "."), cluster.Name, cluster.Namespace)
```

Comment:
I think there's no need to add the path here again. We already have it in the error from the Get function.

Reply:
Yeah, it is a little bit verbose, I'm only replacing the hardcoded path the that was already in the error message with the path from the contract package